keps/0029-operator-dependencies.md

```diff
@@ -0,0 +1,205 @@
+
+### Goals
+
+Dependencies can be a complex topic. This KEP is not trying to boil the dependency ocean but rather limits itself to installation dependencies only.
+
+### Non-Goals
+
+Dependency on an already running `Instance` is a non-goal. It is easy to imagine a situation when a new operator (e.g Kafka) may want to depend on the existing Zookeeper instance. However, such life-cycle dependency presents major challenges e.g. what happens when Zookeeper is removed? What happens when Zookeeper is upgraded and the new version is incompatible with the current Kafka `Instance`? How can compatibility be ensured? This KEP deliberately ignores this area and instead focuses on installation dependencies. Additionally, this KEP does not address output variables or referencing `Instance` resources.
+
+## Proposal
+
+KUDO operators **already have** a mechanism to deal with installation dependencies called [plans, phases, and steps](https://kudo.dev/docs/developing-operators/plans.html#overview) with serial or parallel execution strategy. This mechanism is already powerful enough to express any dependency hierarchy including transitive dependencies (see more about this in [implementation details](#implementation-details)). The core of this proposal is to reuse this mechanism and extend it with the ability to install operators.
+
+### Implementation details
+
+#### Operator Task
+
+A new task kind `Operator` is introduced which extends `operator.yaml` with the ability to install dependencies. Let's take a look at the Kafka+Zookeeper example:
+
+```yaml
+apiVersion: kudo.dev/v1beta1
+name: "kafka"
+operatorVersion: "1.3.1"
+kudoVersion: 0.12.0
+kubernetesVersion: 1.14.8
+appVersion: 2.5.0
+url: https://kafka.apache.org/
+tasks:
+  - name: zookeeper-operator
+    kind: Operator
+    spec:
+        package: zookeeper
+
+  ...
+
+plans:
+  deploy:
+    strategy: serial
+    phases:
+      - name: deploy-zookeeper
+        strategy: serial
+        steps:
+          - name: dummy
+            tasks:
+              - zookeeper-operator
+      - name: deploy-kafka
+        strategy: serial
+        steps:
+          ...
+```
+
+The `zookeeper-operator` task specification is equivalent to `kudo install zookeeper` CLI command which installs the Zookeeper package from the official repo. Here is a complete `Operator` task specification:
+
+```yaml
+tasks:
+- name: demo
+  kind: Operator
+  spec:
+    package: # required, either repo package name, local package folder or an URL to package tarball
+    repo: # optional, name of local repository configuration to use
+    appVersion: # optional, a specific app version in the official repo, defaults to the most recent one
```

Comment:
what if both appVersion and operatorVersion are set and they don't map to existing package? I assume it fails in installation, right? will this be checked before anything is applied to cluster so we don't end up with abandoned OVs?

Reply:
normally, we would be able to just reuse existing `O/OV` (that's what they are for, right?). but this KEP suggests persisting dependency `Instance` parameters in the `OV` which breaks the conceptual separation between `OV` and `I` 😞 Overall, this is IMHO the biggest downsides of this KEP and I haven't come up with a satisfactory solution yet.

Reply:
Here a different solution: 

We pre-create `Instances` on the **client-side** but mark them as _Inactive_ with some annotation e.g. `kudo.de/instance-inactive`. To make things easier IAC would only allow this annotation to be set when `CREATE`ing the instance. This annotation would mark a new life-cycle Instance state where it would effectively be ignored by instance controller.

On the **server-side** we would do the same steps described in this KEP but instead of creating `Instances`, we would activate them (by removing the annotation).

Wdyt?

Reply:
@zen-dog I think @alenkacz is asking what happens if a `kafka` OV says it needs an instance of `zookeeper` version 5, but a package for zookeeper 5 is nowhere to be found in the repo. I think it should just fail at CLI stage, since it's not different from trying to install a non-existent operator `badger`.

I don't understand how reusing existing `O/OV` helps with that?

Reply:
> but a package for zookeeper 5 is nowhere to be found in the repo.

It will fail the same way it fails today for a single operator package with `Error: failed to resolve operator package`